src/xb-builder-node.c

```diff
@@ -265,17 +220,19 @@ xb_builder_node_parse_literal_text(XbBuilderNode *self, const gchar *text, gssiz
 	/* we know this has been pre-fixed */
 	text_len_safe = text_len >= 0 ? (gsize)text_len : strlen(text);
 	if (xb_builder_node_has_flag(self, XB_BUILDER_NODE_FLAG_LITERAL_TEXT))
-		return g_strndup(text, text_len_safe);
+		return xb_arena_strndup(self->arena, text, text_len_safe);
 
 	/* all whitespace? */
 	if (xb_string_isspace(text, text_len_safe))
 		return NULL;
 
 	/* all on one line, no trailing or leading whitespace */
 	if (g_strstr_len(text, text_len, "\n") == NULL)
-		return g_strndup(text, text_len_safe);
+		return xb_arena_strndup(self->arena, text, text_len_safe);
 
 	/* split the text into lines */
+	// Todo: This makes temporary heap allocations, but it's not on the hot path.
```

Comment:
Note: TODO comment here needs addressing (just highlighting it so it doesn’t get forgotten about, I realise this is still work in progress)